internal/wcmatch/wcmatch_test.go

```diff
@@ -0,0 +1,363 @@
+		{"src/file.txt", "/base", "/base/foo/src/file.txt", false},
+		{"src/file.txt", "/base", "/other/foo/src/file.txt", false},
+		{"file.txt", "/base", "/base/foo/src/file.txt", true},
+		{"file.txt", "/base", "/other/foo/src/file.txt", false},
+	}
+
+	for i, tc := range cases {
+		isDir := false
+		// Check if our test path has a trailing slash, indicating it is a directory...
+		if len(tc.path) > 1 && tc.path[len(tc.path)-1] == '/' {
+			// ...if so, remove the trailing slash and set isDir=true
+			isDir = true
+			tc.path = tc.path[:len(tc.path)-1]
+		}
+
+		matcher, err := NewWildcardMatcher(tc.pattern, BaseDir(tc.baseDir))
+
+		if err != nil {
+			t.Errorf("(%v) unexpected error returned for pattern %#v: %v", i, tc.pattern, err)
+		} else {
+			actual := matcher.Match(tc.path, isDir)
+			if actual != tc.expected {
+				t.Errorf("(%v) error matching  pattern %#v with path %#v (case-sensitive): got %v want %v", i, tc.pattern, tc.path, actual, tc.expected)
+			}
+		}
+	}
+}
+
+func TestMatch(t *testing.T) {
+	cases := []wcCase{
+		// Basic
+		{"", "", true, true},
+		{"*", "", true, true},
+		{"**", "", true, true},
+		{"b", "a", false, false},
+		{"*.*", "foo.txt", true, true},
+		{"foo\\.txt", "foo.txt", true, true},
+		{"*/", ".git/", true, true},
+
+		// Sequences
+		{"ab[cd]", "abc", true, true},
+		{"ab[!de]", "abc", true, true},
+		{"[\\\\]", "\\", true, true},
+		{"[!\\\\]", "a", true, true},
+		{"[!\\\\]", "\\", false, false},
+		{"[-abc]", "-", true, true},
+		{"[abc-]", "-", true, true},
+		{"[[]", "[", true, true},
+		{"[a-z]", "q", true, true},
+		{"[a-\\z]", "q", true, true},
+		{"[a-d]", "e", false, false},
+		{"[-abc]", "-", true, true},
+		{"[a\\-c]", "-", true, true},
+		{"[a\\-c]", "b", false, false},
+		{"[[:]", ":", true, true},
+		{"[[:ab]", ":", true, true},
+		{"[[:ab]", "[", true, true},
+		{"[[:ab]", "b", true, true},
+		{"[[:ab]", "c", false, false},
+		{"[![:digit:]ab]", "a", false, false},
+		{"[![:digit:]ab]", "3", false, false},
+		{"[![:digit:]ab]", "c", true, true},
+
+		// Case, general
+		{"abc", "abc", true, true},
+		{"abc", "AbC", false, true},
+		{"AbC", "abc", false, true},
+		{"AbC", "AbC", true, true},
+
+		// Case, sequence
+		{"[A-F]", "C", true, true},
+		{"[A-F]", "c", false, true},
+		{"[a-f]", "c", true, true},
+		{"[a-f]", "C", false, true},
+
+		// Basic wildcard
+		{"ab*", "abcd", true, true},
+		{"ab*cd", "abcd", true, true},
+		{"ab*cd", "abxxxcd", true, true},
+		{"ab***cd", "abcd", true, true},
+		{"ab***cd", "abxxxcd", true, true},
+		{"ab*", "ab/cd", false, false},
+		{"ab**", "ab/cd", false, false},
+		{"?*?", "abc", true, true},
+		{"???*", "abc", true, true},
+		{"*???", "abc", true, true},
+		{"???", "abc", true, true},
+		{"*", "abc", true, true},
+		{"??", "a", false, false},
+
+		// Recursive wildcards
+		{"**", "foo", true, true},
+		{"**", "foo/bar", true, true},
+		{"/a/**/b/foo.txt", "/a/b/c/d/foo.txt", false, false},
```

Comment:
should we add tests here for windows-style paths too with backslashes instead of slashes?

Reply:
... unless of course the paths are always slash-separated, which I don't remember off the top of my head.

Reply:
Backslashes are always escape characters and cannot be used as directory separators from what it seems. So Windows style paths cannot be used in patterns, and since I developed this on Windows I assume that the paths supplied to ignorefs are also never using Windows style directory separators?

Reply:
ah, ok then.